### PR TITLE
adding custom facet support to SparkSQL with tests

### PR DESCRIPTION
**Pull request by swsachith**
Fix for BAM-2128 adding custom facet UDFs

### CODE REVIEW

File: components/analytics-processors/org.wso2.carbon.analytics.spark.core/src/test/java/org/wso2/carbon/analytics/spark/core/udf/UDFTestBase.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+                    time, timeOffset, true));
+        }
+        return result;
+    }
+    /**
+     * generates an random record
+     *
+     * @param tenantId
+     * @param tableName
+     * @param logLevel
```
Comment (madhawa-gunasekara) [inline components/analytics-processors/org.wso2.carbon.analytics.spark.core/src/test/java/org/wso2/carbon/analytics/spark/core/udf/UDFTestBase.java]: Incomplete Java docs

Comment (swsachith) [inline components/analytics-processors/org.wso2.carbon.analytics.spark.core/src/test/java/org/wso2/carbon/analytics/spark/core/udf/UDFTestBase.java]: fixed with the PR : https://github.com/wso2/carbon-analytics/pull/137

File: components/analytics-processors/org.wso2.carbon.analytics.spark.core/src/main/java/org/wso2/carbon/analytics/spark/core/udf/facets/FacetUDF.java [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+ * under the License.
+ */
+
+package org.wso2.carbon.analytics.spark.core.udf.facets;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * represents the Facet UDF
```
Comment (lafernando) [inline components/analytics-processors/org.wso2.carbon.analytics.spark.core/src/main/java/org/wso2/carbon/analytics/spark/core/udf/facets/FacetUDF.java]: A line should starts with a capital letter and should ends with a period.

Comment (swsachith) [inline components/analytics-processors/org.wso2.carbon.analytics.spark.core/src/main/java/org/wso2/carbon/analytics/spark/core/udf/facets/FacetUDF.java]: This has been fixed in https://github.com/wso2/carbon-analytics/pull/137

**Pull request merged**